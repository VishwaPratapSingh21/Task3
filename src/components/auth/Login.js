import React, {Component} from 'react';
import axiosInstance from '../Apicalls';
import {notify} from 'react-notify-toast';
import {Link, Redirect} from 'react-router-dom';
import PasswordResetEmail from './ResetEmail';
import PropTypes from 'prop-types'

/**
 * Component for handling login.
 * @author [Hadijah kyampeire](https://github.com/hadijahkyampeire/Yummy_Reactjs_frontend)
 */
class Login extends Component {
    // initial states
    state = {
        password_field: '',
        email_field: '',
        error:null,
    }

    static propTypes = {
        login: PropTypes.func.isRequired
    }
    // handles changes and updates state
    handleInputChange = (event) => {
        const {name, value} = event.target;
        this.setState({[name]: value});
    }
    // function to make an API call on submit
    handleLogin = (event) => {
        const {email_field: email, password_field: password} = this.state;

        event.preventDefault();
        axiosInstance
            .post('auth/login', {email, password})
            .then(response => {
                localStorage.setItem('accessToken', response.data.access_token);
                localStorage.setItem('userEmail', response.data.user_email);
                localStorage.setItem('username', response.data.username);
                localStorage.setItem('identity', response.data.Id);
                this
                    .props
                    .login();
                notify.show(response.data.message, 'success', 4000);
            })
            // error handling
            .catch(error => {
                if (error.response) {
                    this.setState({error: error.response.data.message})
                    // notify.show(error.response.data.message, 'error', 3000)
                } else if (error.request) {
                    notify.show("Request not made", 'error', 3000)
                }
            });
    }
    render() {
        const {password_field, email_field} = this.state;
        const {from} = this.props.location.state || {
            from: {
                pathname: '/dashboard'
            }
        }

        if (this.props.loggedIn) {
            return (<Redirect to={from}/>)
        }
        
        return (
            <div className="background">
                <div id="logbar">
                    <div className="form-group">
                    <i className="fa fa-user-circle fa-5x" style={{color:"#26A69A"}}></i>
                        <div>
                            LOGIN
                        </div>
                        <form onSubmit={this.handleLogin}>
                        {this.state.error ? (
                    <div className="alert alert-danger">{this.state.error}
                     <button
                    type="button"
                    class="close"
                    data-dismiss="alert"
                    aria-label="Close"
                  >
                    <span aria-hidden="true">&times;</span>
                  </button></div>
                  ) : (
                    ""
                  )}
                            <div className="input-group">
                                <div className="input-group-addon pr-4" ><i className='fa fa-envelope'/></div>
                                <input
                                    type="email"
                                    id="emailsignup"
                                    name="email_field"
                                    aria-describedby="email"
                                    className="form-control"
                                    placeholder="example@gmail.com"
                                    onChange={this.handleInputChange}
                                    value={email_field}/>
                            </div>
                            <br/>
                            <div className="input-group">
                                <div className="input-group-addon pr-4" ><i className='fa fa-lock'/></div>
                                <input
                                    type="password"
                                    name="password_field"
                                    className="form-control"
                                    onChange={this.handleInputChange}
                                    value={password_field}
                                    placeholder="Password"/>
                            </div>
                            <br/>

                            <button name="signin" className="btn btn-default btn-lg" id="signin" type="submit">
                               
                                    LOGIN
                            </button>
                        </form>
                        <div className="change_link">
                            Not a member yet ?
                            <Link to="/register">
                                < h4 className="toregister">Register Here</h4>
                            </Link>
                        </div>
                    </div>
                    <a href="#" data-toggle="modal" data-target="#forgot_password">ForgotPassword?</a>
                    <PasswordResetEmail/>
                </div>
                <footer className="footer">
                    <p>&copy; Hadijah 2018</p>
                </footer>
            </div>
        );
    }
}

export default Login;
