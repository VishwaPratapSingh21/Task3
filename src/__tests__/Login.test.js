import React from 'react';
import { shallow} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Login from '../components/auth/Login';

describe('Login component', () => {
  const login = jest.fn();
  const wrapper = shallow(<Login login={login} loggedIn={false} location={{}} />);
  const preventDefault = jest.fn();

  it('renders properly without crashing', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('it renders state initially', () => {
    expect(wrapper.state().email_field).toEqual('');
    expect(wrapper.state().password_field).toEqual('');
  });

  it('has the correct form fields', () => {
    expect(wrapper.find('#emailsignup')).toHaveLength(1);
    expect(wrapper.find('[name="password_field"]')).toHaveLength(1);
    expect(wrapper.find('#signin')).toHaveLength(1);
  });

  it('Form fields update when state changes', () => {
    wrapper.setState({ password_field: 'hi', email_field: 'mail' });
    expect(wrapper.find('#emailsignup').props().value).toEqual('mail');
    expect(wrapper.find('[name="password_field"]').props().value).toEqual('hi');
  });

  it('renders component in div', () => {
    expect(wrapper.find('div').length).toBe(9);
  });

  it('renders component in a form', () => {
    expect(wrapper.find('form').length).toBe(1);
  });

  it('renders inputs', () => {
    expect(wrapper.find('input').length).toBe(2);
  });

  it('renders buttons', () => {
    expect(wrapper.find('button').length).toBe(1);
  });

  it('it has br', () => {
    expect(wrapper.find('br').length).toBe(2);
  });

  it('submits data', () => {
    wrapper.find('form').simulate('submit', { preventDefault });
    expect(preventDefault).toBeCalled();
  });
});
