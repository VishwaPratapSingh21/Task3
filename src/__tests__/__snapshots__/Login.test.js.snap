// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component renders properly without crashing 1`] = `
<div
  className="background"
>
  <div
    id="logbar"
  >
    <div
      className="form-group"
    >
      <i
        className="fa fa-user-circle fa-5x"
        style={
          Object {
            "color": "#26A69A",
          }
        }
      />
      <div>
        LOGIN
      </div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="input-group"
        >
          <div
            className="input-group-addon pr-4"
          >
            <i
              className="fa fa-envelope"
            />
          </div>
          <input
            aria-describedby="email"
            className="form-control"
            id="emailsignup"
            name="email_field"
            onChange={[Function]}
            placeholder="example@gmail.com"
            type="email"
            value=""
          />
        </div>
        <br />
        <div
          className="input-group"
        >
          <div
            className="input-group-addon pr-4"
          >
            <i
              className="fa fa-lock"
            />
          </div>
          <input
            className="form-control"
            name="password_field"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <br />
        <button
          className="btn btn-default btn-lg"
          id="signin"
          name="signin"
          type="submit"
        >
          LOGIN
        </button>
      </form>
      <div
        className="change_link"
      >
        Not a member yet ?
        <Link
          replace={false}
          to="/register"
        >
          <h4
            className="toregister"
          >
            Register Here
          </h4>
        </Link>
      </div>
    </div>
    <a
      data-target="#forgot_password"
      data-toggle="modal"
      href="#"
    >
      ForgotPassword?
    </a>
    <PasswordResetEmail />
  </div>
  <footer
    className="footer"
  >
    <p>
      Â© Hadijah 2018
    </p>
  </footer>
</div>
`;
